File: src/main/java/com/bma/problemsolving/leetcode/cpp/letters_combination_of_a_phone_number.cpp
#include<iostream>
#include<string>
#include<vector>
#include<unordered_map>

using namespace std;

void findAll(string combinations, string& digits, int digit_index, unordered_map<char,string> telephone_num_chars, vector<string> &output) {
    if (digit_index == digits.length()) {
        output.push_back(combinations);
        return;
    }

    for (char eachChar: telephone_num_chars[digits[digit_index]]) {
        combinations[digit_index] = eachChar;
        findAll(combinations, digits, digit_index + 1, telephone_num_chars, output);
    }
    
}

int main() {
    string digits;
    cin  >> digits;

    if (digits.empty()) return {};

    unordered_map<char,string> telephone_num_chars;
    telephone_num_chars.insert({'2', "abc"});
    telephone_num_chars.insert({'3', "def"});
    telephone_num_chars.insert({'4', "ghi"});
    telephone_num_chars.insert({'5', "jkl"});
    telephone_num_chars.insert({'6', "mno"});
    telephone_num_chars.insert({'7', "pqrs"});
    telephone_num_chars.insert({'8', "tuv"});
    telephone_num_chars.insert({'9', "wxyz"});

   
    vector<string> output;
    string combinations(digits.length(), ' ');
    findAll(combinations, digits, 0, telephone_num_chars, output);

    for (string combination: output) {
        cout << combination  << ",";
    }
}

File: src/main/java/com/bma/problemsolving/leetcode/java/backtracking/LettersCombinationOfAPhoneNumber.java
package com.bma.problemsolving.leetcode.java.backtracking;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * @author Varun Shrivastava
 * @github www.github.com/vslala
 * @date 22/05/2025
 */
class LettersCombinationOfAPhoneNumber {
    private Map<Character, String> letters = Map.of(
            '2',
            "abc",
            '3',
            "def",
            '4',
            "ghi",
            '5',
            "jkl",
            '6',
            "mno",
            '7',
            "pqrs",
            '8',
            "tuv",
            '9',
            "wxyz"
    );
    private List<String> combinations;
    private String digits;
    public List<String> letterCombinations(String digits) {
        this.digits = digits;
        this.combinations = new ArrayList<>();
        backtrack(0, new StringBuilder());
        return this.combinations;
    }

    private void backtrack(int index, StringBuilder path) {
        if (path.length() == this.digits.length()) {
            this.combinations.add(path.toString());
            return;
        }

        String possibleLetters = letters.get(digits.charAt(index));
        for (char letter: possibleLetters.toCharArray()) {
            path.append(letter);
            backtrack(index + 1, path);
            path.deleteCharAt(path.length() - 1);
        }
    }
}
